### PR TITLE
Roe 2196 add authcodeservice

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link

https://companieshouse.atlassian.net/browse/ROE-2196

### Change description

Added required service code to call the sdk and retrieve the email for the specified company from the oracle query api.

### Work checklist

- [x] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -58,6 +58,21 @@
       <artifactId>structured-logging</artifactId>
       <version>${structured-logging.version}</version>
     </dependency>
+    <dependency>
+      <groupId>uk.gov.companieshouse</groupId>
+      <artifactId>api-sdk-manager-java-library</artifactId>
+      <version>1.0.4</version>
+    </dependency>
+    <dependency>
+      <groupId>uk.gov.companieshouse</groupId>
```
Comment (mwestacott) [inline pom.xml:67]: Just wondering, if we're hitting the private API to get the e-mail, do we need to include this public SDK too? Quick check - without it, does the call still work?

File: src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java [hunk 2]
```diff
@@ -16,13 +19,24 @@
...
                                             @PathVariable String companyNumber) {
-
         DataMap dataMap = new DataMap.Builder().companyNumber(companyNumber).build();
         ApiLogger.infoContext(requestId,"Request received for auth code email", dataMap.getLogMap());
 
+        try {
+            authCodeNotificationService.sendAuthCodeEmail(requestId, companyNumber);
+        } catch (ServiceException e) {
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (Exception e) {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java:35]: Still not convinced we need this catch block for Exception (even though it's there in OE API). If it's removed, does code still compile, pass tests and run?

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationService.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @Autowired
+    private PrivateDataRetrievalService privateDataRetrievalService;
+
+    public void sendAuthCodeEmail(String requestId, String companyNumber) throws ServiceException {
+        DataMap dataMap = new DataMap.Builder().companyNumber(companyNumber).build();
+        ApiLogger.infoContext(requestId, "Send auth code email invoked", dataMap.getLogMap());
+
+        String email = getOverseasEntityEmail(requestId, companyNumber, dataMap.getLogMap());
+
+        // 2. send email
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationService.java:24]: Could make this a TODO comment. At the moment it looks like something has forgotten to be implemented.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationService.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    private String getOverseasEntityEmail(String requestId, String companyNumber, Map<String, Object> logMap) throws ServiceException {
+        String email = privateDataRetrievalService.getOverseasEntityData(requestId, companyNumber).getEmail();
+
+        if (Strings.isBlank(email)) {
+            ServiceException e = new ServiceException("Null or empty email found");
+            ApiLogger.errorContext(requestId, "Failed to retrieve complete email", e, logMap);
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationService.java:32]: Suggest changing this message to 'Failed to retrieve a valid email'.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import uk.gov.companieshouse.authcodenotification.exception.ServiceException;
+import uk.gov.companieshouse.authcodenotification.utils.ApiLogger;
+import uk.gov.companieshouse.logging.util.DataMap;
+
+import java.io.IOException;
+
+@Service
+public class PrivateDataRetrievalService {
+
+    private static final String OVERSEAS_ENTITY_URI_SECTION = "/overseas-entity/";
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:17]: Suggest we making this a 'formattable' string (with %s in for company number) and use the string formatter to populate it, rather than do all the string concatenation.

File: src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java [hunk 6]
```diff
@@ -1,16 +1,40 @@
...
-        AuthCodeNotificationController controller = new AuthCodeNotificationController();
-        ResponseEntity<Object> responseEntity = controller.sendEmail("abc","012345657");
+    @Test
+    void testSendEmailReturnsSuccess() {
+        ResponseEntity<Object> responseEntity = controller.sendEmail(REQUEST_ID ,COMPANY_NUMBER);
         assertEquals( 200, responseEntity.getStatusCode().value() );
     }
+
+    @Test
+    void testSendEmailException() throws ServiceException{
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java:35]: Suggest renaming this to 'testSendEmailReturnsInternalServerErrorWhenServiceCallFails'.

File: src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    void testSuccessfulRetrieveAndSend() throws ServiceException {
+        OverseasEntityDataApi overseasEntityDataApi = new OverseasEntityDataApi();
+        overseasEntityDataApi.setEmail(TEST_EMAIL);
+        when(privateDataRetrievalService.getOverseasEntityData(REQUEST_ID, COMPANY_NUMBER)).thenReturn(overseasEntityDataApi);
+        authCodeNotificationService.sendAuthCodeEmail(REQUEST_ID, COMPANY_NUMBER);
+        verify(privateDataRetrievalService, times(1)).getOverseasEntityData(REQUEST_ID, COMPANY_NUMBER);
+    }
+
+    @Test
+    void testBlankEmailExceptionThrown() throws ServiceException {
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java:41]: 'testExceptionThrownWhenEmailIsBlankOrNotFound'

File: src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java:35]: Tests still pass.

File: pom.xml (hunk 1 above)
Comment (markpit) [reply pom.xml:67]: we had to add it for some imports, think it was some exceptions being thrown from private sdk

File: src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java (hunk 7 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java]: Method name changed.

File: src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java (hunk 6 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java]: Method name changed.

File: src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java (hunk 2 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java:35]: Suggests it can go. I'd just test with the call in Docker too though before removing.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 5 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:17]: yes, can remove the constant and just use /overseas-entity/ in the string.format.

File: src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java (hunk 6 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationControllerTest.java]: Thanks

File: src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java (hunk 7 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java]: Thanks for the change but think you might have done that based on an older version of the comment. Name needs to be specific to the test you're checking - in this case it's a null email being set but we should also have a test for a blank email being set (or a a string containing just whitespace).

Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/authcodenotification/service/AuthCodeNotificationServiceTest.java]: Separated the tests

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 5 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java]: Kept the constant but made it formatable.

File: src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/controller/AuthCodeNotificationController.java]: Removed and tested in docker

**Review by mwestacott — DISMISSED** [review]

**Review by mwestacott — APPROVED** [review]
